cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamCoordinator.java

```diff
@@ -0,0 +1,203 @@
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link StreamCoordinator} uses ZooKeeper to implementation coordination needed for stream. It also uses a
+ * {@link ResourceCoordinator} to elect each handler as the leader of a set of streams.
+ */
+@Singleton
+public final class DistributedStreamCoordinator extends AbstractStreamCoordinator {
+  private static final Logger LOG = LoggerFactory.getLogger(DistributedStreamCoordinator.class);
+
+  public static final String STREAMS_COORDINATOR = "streams.coordinator";
+  private static final String STREAMS_RESOURCE = "streams";
+
+  private ZKClient zkClient;
+
+  private final DiscoveryServiceClient discoveryServiceClient;
+  private final StreamMetaStore streamMetaStore;
+  private final ResourceCoordinatorClient streamsResourceCoordinatorClient;
+
+  private LeaderElection leaderElection;
+  private ResourceCoordinator streamsResourceCoordinator;
+  private Discoverable handlerDiscoverable;
+  private Cancellable handlerSubscription;
+
+  @Inject
+  public DistributedStreamCoordinator(StreamAdmin streamAdmin, ZKClient zkClient,
+                                      DiscoveryServiceClient discoveryServiceClient,
+                                      StreamMetaStore streamMetaStore) {
+    super(streamAdmin);
+    this.zkClient = zkClient;
+    this.discoveryServiceClient = discoveryServiceClient;
+    this.streamMetaStore = streamMetaStore;
+    this.streamsResourceCoordinatorClient = new ResourceCoordinatorClient(zkClient);
+    this.handlerDiscoverable = null;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    Preconditions.checkNotNull(handlerDiscoverable, "Stream Handler discoverable has not been set");
+    streamsResourceCoordinatorClient.startAndWait();
+
+    // Start the resource coordinator that will map Streams to Stream handlers
+    leaderElection = new LeaderElection(
+      zkClient, "/election/" + STREAMS_COORDINATOR, new ElectionHandler() {
+      @Override
+      public void leader() {
+        LOG.info("Became Stream handler leader. Starting resource coordinator.");
+        streamsResourceCoordinator = new ResourceCoordinator(zkClient, discoveryServiceClient,
+                                                             new BalancedAssignmentStrategy());
+        streamsResourceCoordinator.startAndWait();
+
+        try {
+          // Create one requirement for the resource coordinator for all the streams.
+          // One stream is identified by one partition
+          ResourceRequirement.Builder builder = ResourceRequirement.builder(STREAMS_RESOURCE);
+          for (StreamSpecification spec : streamMetaStore.listStreams()) {
+            LOG.debug("Adding {} stream as a resource to the coordinator to manager streams leaders.", spec.getName());
+            builder.addPartition(new ResourceRequirement.Partition(spec.getName(), 1));
+          }
+          streamsResourceCoordinatorClient.submitRequirement(builder.build()).get();
```

Comment:
Not necessarily since there can be race. Consider there is a switch in leader due to ZK connection timeout (hence this `leader()` will be executed). So the following happens:

1. The leader scan the meta store to get a list of streams
2. There is another handler instance (another process) that has a stream created and updated the requirement.
3. The leader set the requirement, it won't have the newly created stream as created by another handler instance.